### PR TITLE
Add local APC cache to /area

### PR DESCRIPTION
**Pull request by esainane**
## About The Pull Request

`for(var/obj/machinery/power/apc/APC in GLOB.apcs_list)` in `get_apc` makes chuck norris shed a tear

## Why It's Good For The Game

tears are bad

## Changelog

:cl:
code: /area now retains a local reference to the APC in the area, if any
/:cl:

### CODE REVIEW

File: code/modules/power/apc.dm [hunk 1]
```diff
@@ -220,6 +220,12 @@
 	..()
 	GLOB.apcs_list += src
 
+	var/area/local_area = get_area(src)
+	if (local_area)
+		if (local_area.apc)
+			WARNING("Duplicate APC created at [AREACOORD(src)]")
+		local_area.apc = src
```
Comment (LemonInTheDark) [inline code/modules/power/apc.dm:227]: What happens if an apc moves?

Comment (esainane) [reply code/modules/power/apc.dm:227]: I suppose that can happen near a Singularity? I can listen for COMSIG_ENTER_AREA and COMSIG_EXIT_AREA if need be, but I have no idea if the existing behavior is coherent there

**Comment by esainane** [comment]
I have no idea why we set `area` both in `New` and `Initialize` and using the one from `New` breaks everything, but at this point I'm afraid to find out

**Comment by esainane** [comment]
nothing touches `area` after `Initialize` so is presumably valid from `Initialize` until `Destroy` (and doesn't handle area movement, if any happens)

File: code/modules/power/apc.dm (hunk 1 above)
Comment (Mothblocks) [reply code/modules/power/apc.dm]: @LemonInTheDark APCs moving happens so infrequently that we should just completely disconnect them when they do. APC code is hacked together enough as is from what I know, we don't even support two in the same area.

Comment (LemonInTheDark) [reply code/modules/power/apc.dm]: that would be fine. just so long as the refs don't hang

Comment (esainane) [reply code/modules/power/apc.dm]: It's cleaned up in Destroy, which is also where the reference in the GLOB apc list is cleared, so the lifecycle should remain unchanged

A duplicate APC will clear an area's existing reference to an APC (and print a warning), but removing a reference early shouldn't cause it to hang around longer than clearing the reference later

**Pull request merged**